### PR TITLE
First commit for the Multiple Interpreters

### PR DESCRIPTION
**Pull request by esakellari**
First commit for the Multiple Interpreters.
--------------------------------------------------------
--------------------------------------------------------

ASTImportSource.cpp, ASTImportSource.h
--------------------------------------------------------------
The class that implements the importing functionalities for the multiple Interpreters.
It derives from the clang class ExternalASTSource, and overrides the function  FindExternalVisibleDeclsByname which does the search and importing of the ASTNodes by  calling the clang::ASTImporter.
This class must be created when setting up the second Interpreter and “passed” to its Translation Unit with the function setExternalSource.

Symbol Resolution from the execution engine of the second interpreter.
---------------------------------------------------------------------------------------------------
In order for the second Interpreter to find and execute the correct functions that are defined in the first Interpreter, but are called from the second, the following changes have been made in the classes of IncrementalExecutor and Interpreter.

(IncrementalExecutor.cpp, IncrementalExecutor.h, Interpreter.cpp, Interpreter.h)

-Added a pointer to an 'external' IncrementalExecutor in IncrementalExecutor class, and a corresponding function to set the pointer (IncrementalExecutor::setExternalIncrementalExecutor).
    
-Added in Interpreter a getIncrementalExecutor() function to get a pointer to its IncrementalExecutor.
    
-Added in Interpreter a function Interpreter::setExternalIncrementalExecutor which calls IncrementalExecutor's setExternalIncrementalExecutor function.
    
 -Edited IncrementalExecutor::NotifyLazyFunctionCreators to search for the address of a missing symbol in the external IncrementalExecutor that has been set. If it doesn't exist in the external IncrementalExecutor either, it goes to HandleMissingFunction.

Interpreter constructor overloading 
-------------------------------------------------
The overloading constructor of the Interpreter takes as argument the pointer to the master Interpreter. This is used for the initialization of the IncrementalParser  and to avoid the set up or the runtime environment in the constructor. This is done to reduce the memory consumption during the creation and set up of the second Interpreter. 

IncrementalParser.cpp, IncrementalParser.h
-------------------------------------------------------------
The function Initialize() of the IncrementalParser takes a bool argument to know whether it is called from a “master” Interpreter or from a “slave” Interpreter.
If it is a slave Interpreter, it avoids to include <new>, and in turn it reduces the memory consumption for the creation of the second Interpreter. 

Small change in /clang/lib/Sema/SemaOpenMP.cpp
-----------------------------------------------------------------------
This small change has been done in order to improve the memory consumption of the multiple interpreters.

### CODE REVIEW

File: interpreter/cling/include/cling/Interpreter/Interpreter.h [hunk 1]
```diff
@@ -293,6 +304,10 @@ namespace cling {
 
     llvm::LLVMContext* getLLVMContext() { return m_LLVMContext.get(); }
 
+    IncrementalExecutor* getIncrementalExecutor() { return m_Executor.get(); }
+
+    void setExternalIncrementalExecutor(IncrementalExecutor* extIncr);
```
Comment (karies) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: What do we need this getter and setter for? They seem to be unused, and I don't see a need for them.

File: interpreter/cling/include/cling/Interpreter/Interpreter.h [hunk 2]
```diff
@@ -282,6 +282,17 @@ namespace cling {
...
     Interpreter(int argc, const char* const *argv, const char* llvmdir = 0, bool noRuntime = false);
+
+    ///\brief Constructor for Interpreter.
+    ///
+    ///\param[in] argc - no. of args.
+    ///\param[in] argv - arguments passed when driver is invoked.
+    ///\param[in] llvmdir - ???
+    ///\param[in] noRuntime - flag to control the presence of runtime universe
+    ///\param[in] originalInterpreter - the master interpreter of this interpreter
+    Interpreter(Interpreter *originalInterpreter,int argc, const char* const *argv,
```
Comment (karies) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: Can we pass in an `Interpreter&` instead of a pointer? This is required to be non null, passing a reference signals that.

File: interpreter/cling/include/cling/Interpreter/Interpreter.h (hunk 1 above)
Comment (esakellari) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: Yes, this is because they are used for the set-up of the second 
interpreter.

File: interpreter/cling/include/cling/Utils/ASTImportSource.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  class IntegerLiteral;
+  class NamedDecl;
+  class NamespaceDecl;
+  class NestedNameSpecifier;
+  class QualType;
+  class Sema;
+  class TagDecl;
+  class TemplateDecl;
+  class Type;
+  class TypedefNameDecl;
```
Comment (karies) [inline interpreter/cling/include/cling/Utils/ASTImportSource.h]: Can you re-check the list of forward decls? That seems to be more than the header file needs...

File: interpreter/cling/include/cling/Utils/ASTImportSource.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+namespace cling {
+  namespace utils {
+
+    class ASTImportSource : public clang::ExternalASTSource {
```
Comment (karies) [inline interpreter/cling/include/cling/Utils/ASTImportSource.h]: Can you mark all virtual functions that your class re-implements with `override`?

File: interpreter/cling/include/cling/Utils/ASTImportSource.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+      private:
+        cling::Interpreter *m_first_Interp;
```
Comment (karies) [inline interpreter/cling/include/cling/Utils/ASTImportSource.h]: Isn't the first interpreter const? Does it work if you keep a const pointer?

File: interpreter/cling/include/cling/Utils/ASTImportSource.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        cling::Interpreter *m_second_Interp;
+        const clang::TranslationUnitDecl *m_translationUnitI1;
+        const clang::TranslationUnitDecl *m_translationUnitI2;
```
Comment (karies) [inline interpreter/cling/include/cling/Utils/ASTImportSource.h]: You don't actually need that, right? Can we just keep the ASTContext for the second interpreter?

File: interpreter/cling/include/cling/Utils/ASTImportSource.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+        clang::Sema *m_Sema;
+        /* A map for the DeclContext-to-DeclContext correspondence
```
Comment (karies) [inline interpreter/cling/include/cling/Utils/ASTImportSource.h]: Can you make that comment Doxygen style? Can you state that key is second interpreter and value is first interpreter (if that's what it is)? That's really essential information for this map - the fact that it's a map of decl context to decl context is clearly visible from the type - but you ought to explain what it's meaning is. You should probably even state why you have it. All of that should still only be one our two sentences :-) This is the most important member of your class!

File: interpreter/cling/include/cling/Utils/ASTImportSource.h [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+         * with DeclContexts pointers. */
+        std::map<const clang::DeclContext *, clang::DeclContext *> m_DeclContexts_map;
+        /* A map for all the imported Decls (Contexts). */
```
Comment (karies) [inline interpreter/cling/include/cling/Utils/ASTImportSource.h]: Ditto: what is key, what is value (not their type but their meaning!) And what is this map used for?

File: interpreter/cling/include/cling/Interpreter/Interpreter.h (hunk 1 above)
Comment (karies) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: Strange, I cannot find a call in this pull request?

File: interpreter/cling/include/cling/Utils/ASTImportSource.h [hunk 9]
```diff
@@ -0,0 +1,88 @@
+#ifndef CLING_ASTIMPORTSOURCE_H
```
Comment (karies) [inline interpreter/cling/include/cling/Utils/ASTImportSource.h]: Can you add the usual file header? Copy and adjust it from any other cling header.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp [hunk 10]
```diff
@@ -171,8 +177,12 @@ IncrementalExecutor::NotifyLazyFunctionCreators(const std::string& mangled_name)
     if (ret)
       return ret;
   }
+  llvm::StringRef name(mangled_name);
+  void *address = nullptr;
+  if(m_externalIncrementalExecutor)
+   address = m_externalIncrementalExecutor->getAddressOfGlobal(name);
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: This shouldn't happen here but in the IncrementalJIT which does the symbol lookup. Reason is that it might compile (codegen) a function which in turn needs a symbol - which should be resolved from the previous interpreter. So chaining the IncrementalJITs seems to be the right approach. They already have a name -> symbol function that we'll need to extend.

File: interpreter/cling/lib/Interpreter/IncrementalParser.cpp [hunk 11]
```diff
@@ -202,7 +202,7 @@ namespace cling {
 
   void
   IncrementalParser::Initialize(llvm::SmallVectorImpl<ParseResultTransaction>&
-                                result) {
+                                result, bool slaveInterp) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/IncrementalParser.cpp]: Can we use `subInterp` or `derivedInterp` or `extendingInterp`?

File: interpreter/cling/lib/Interpreter/IncrementalParser.cpp [hunk 12]
```diff
@@ -247,14 +247,16 @@ namespace cling {
...
-    if (m_CI->getLangOpts().CPlusPlus) {
-      // <new> is needed by the ValuePrinter so it's a good thing to include it.
-      // We need to include it to determine the version number of the standard
-      // library implementation.
-      ParseInternal("#include <new>");
-      // That's really C++ ABI compatibility. C has other problems ;-)
-      CheckABICompatibility(m_CI.get());
-    }
+    // If I am the master Interpreter, only then do the #include <new>
+    if (!slaveInterp)
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/IncrementalParser.cpp]: Can you add `{` for this `if`, or combing this condition with the next one (`CPlusPlus`)?

File: interpreter/cling/lib/Interpreter/Interpreter.cpp [hunk 13]
```diff
@@ -224,6 +224,72 @@ namespace cling {
     setCallbacks(std::move(AutoLoadCB));
   }
 
+  //Overloading the Interpreter constructor by passing one more argument, the pointer
+  //to the "master" Interpreter.
+  Interpreter::Interpreter(Interpreter *originalInterpreter, int argc, const char* const *argv,
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/Interpreter.cpp]: Instead of duplicating another Interpreter c'tor, can we factor most of the code out? Can we maybe have the existing one call this one or vice versa? Check "constructor delegation", see e.g. https://msdn.microsoft.com/en-us/library/dn387583.aspx.

File: interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp [hunk 14]
```diff
@@ -420,7 +420,7 @@ namespace {
     R.clear();
     R.setLookupName(&m_Context->Idents.get("copyArray"));
     m_Sema->LookupQualifiedName(R, NSD);
-    assert(!R.empty() && "Cannot find cling::runtime::internal::copyArray");
+    //assert(!R.empty() && "Cannot find cling::runtime::internal::copyArray");
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp]: Can you add a `//FIXME` with some text why this is commented, and when it should be re-enabled? Can we make this conditional, e.g. "only assert if we have no parent interpreter"?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+                                     cling::Interpreter *interpreter_second) :
+      m_first_Interp(interpreter_first), m_second_Interp(interpreter_second) {
+
+      m_translationUnitI1 = m_first_Interp->getCI()->getASTContext().getTranslationUnitDecl();
+      m_translationUnitI2 = m_second_Interp->getCI()->getASTContext().getTranslationUnitDecl();
+
+    }
+
+    void ASTImportSource::InitializeSema(Sema &S) {
+      m_Sema = &S;
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Make inline?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    void ASTImportSource::ForgetSema() {
+      m_Sema = nullptr;
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Make inline?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+          }
+        } else {
+          Decl *declFrom = llvm::cast<Decl>(*lookup_result.data());
+          ImportDecl(declFrom, importer, Name, declNameI1, DC);
+        }
+      }
+      return true;
+    }
+
+    bool ASTImportSource::FindExternalVisibleDeclsByName(const DeclContext *DC,
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Can you add `1` and `2` or `parent` and `child` as prefix to all `Decl` / `DeclContext`, to signal which interpreter they are from? This makes your code much more readable. E.g. is this `DC` from the child or the parent interpreter?

File: interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp [hunk 18]
```diff
@@ -420,7 +421,10 @@ namespace {
     R.clear();
     R.setLookupName(&m_Context->Idents.get("copyArray"));
     m_Sema->LookupQualifiedName(R, NSD);
-    assert(!R.empty() && "Cannot find cling::runtime::internal::copyArray");
+    /* If I am not being called from a child Interpreter, only then do
+     * the assert. */
+    if (!m_isChildInterpreter)
+     assert(!R.empty() && "Cannot find cling::runtime::internal::copyArray");
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp]: Indentation seems wrong (2 chars)

File: interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp [hunk 19]
```diff
@@ -420,7 +421,10 @@ namespace {
     R.clear();
     R.setLookupName(&m_Context->Idents.get("copyArray"));
     m_Sema->LookupQualifiedName(R, NSD);
-    assert(!R.empty() && "Cannot find cling::runtime::internal::copyArray");
+    /* If I am not being called from a child Interpreter, only then do
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp]: Well - this is a temporary hack until we can import copyArray. And once we can import it, this member and the assert-protection should go away.

Could you clarify that in the comment, please?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 20]
```diff
@@ -0,0 +1,189 @@
+#include "cling/Utils/ASTImportSource.h"
+
+using namespace clang;
+
+namespace cling {
+  namespace utils {
+    ASTImportSource::ASTImportSource(cling::Interpreter *parent_interpreter,
+                                     cling::Interpreter *child_interpreter) :
+      m_parent_Interp(parent_interpreter), m_child_Interp(child_interpreter) {
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Can we have this inlined, please?

File: interpreter/cling/include/cling/Interpreter/Interpreter.h [hunk 21]
```diff
@@ -281,7 +281,22 @@ namespace cling {
     ///\param[in] argv - arguments passed when driver is invoked.
     ///\param[in] llvmdir - ???
     ///\param[in] noRuntime - flag to control the presence of runtime universe
-    Interpreter(int argc, const char* const *argv, const char* llvmdir = 0, bool noRuntime = false);
+    ///\param[in] isChildInterp - flag to control if this is a child interpreter or not.
+    ///
+    Interpreter(int argc, const char* const *argv,
+                const char* llvmdir = 0, bool noRuntime = false, bool isChildInterpreter = false);
```
Comment (karies) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: Can we have the union of the two interfaces in a non-public (likely private) function, instead of adding a parameter to a public interface that doesn't make sense when called by users? Or do you think there's a use case for having this constructor called with `isChildInterpreter = true` in user code? If not let's move the common functionality out, and keep the public constructors with as few parameters as possible. (There are already too many...)

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp [hunk 22]
```diff
@@ -171,8 +178,12 @@ IncrementalExecutor::NotifyLazyFunctionCreators(const std::string& mangled_name)
     if (ret)
       return ret;
   }
+  llvm::StringRef name(mangled_name);
+  void *address = nullptr;
+  if(m_externalIncrementalExecutor)
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: space after `if`.

File: interpreter/cling/lib/Interpreter/Interpreter.cpp [hunk 23]
```diff
@@ -224,6 +224,17 @@ namespace cling {
     setCallbacks(std::move(AutoLoadCB));
   }
 
+  ///\brief Constructor for the child Interpreter.
+  /// Passing the parent Interpreter as an argument.
+  ///
+  Interpreter::Interpreter(Interpreter &parentInterpreter, int argc, const char* const *argv,
+                           const char* llvmdir /*= 0*/, bool noRuntime, bool isChildInterpreter) :
+    Interpreter::Interpreter(argc, argv, llvmdir /*= 0*/, noRuntime, isChildInterpreter) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/Interpreter.cpp]: Very nice, thanks!

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    void ASTImportSource::ImportDecl(Decl *declToImport,
+                                     ASTImporter &importer,
+                                     DeclarationName &childDeclName,
+                                     DeclarationName &parentDeclName,
+                                     const DeclContext *childCurrentDeclContext) {
+
+      Decl *importedDecl = nullptr;
+      /* Don't do the import if we have a Function Template.
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: We generally prefer C++ comments, throughout, except for ROOT's doxygen class doc. Can you convert all comments to `//`?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+       * Not supported by clang. */
+      if (!(declToImport->isFunctionOrFunctionTemplate() && declToImport->isTemplateDecl()))
+        importedDecl = importer.Import(declToImport);
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: what about `if (Decl *importedDecl = ...` and then move the block below here?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+
+      if (importedDecl) {
+        /* Import this Decl in the map we own. Not used for now
+         * m_Decls_map[Name.getAsString()] = std::make_pair(importedDecl, declToImport); */
+        std::vector < NamedDecl * > declVector;
+        declVector.push_back((NamedDecl *) importedDecl);
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: How certain are you that this is a `NamedDecl`?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+      Decl *importedDecl = nullptr;
+      /* Don't do the import if we have a Function Template.
+       * Not supported by clang. */
+      if (!(declToImport->isFunctionOrFunctionTemplate() && declToImport->isTemplateDecl()))
+        importedDecl = importer.Import(declToImport);
+
+      if (importedDecl) {
+        /* Import this Decl in the map we own. Not used for now
+         * m_Decls_map[Name.getAsString()] = std::make_pair(importedDecl, declToImport); */
+        std::vector < NamedDecl * > declVector;
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: You could use `std::vector < NamedDecl * > declVector{importedNamedDecl};` instead of the runtime `push_back()`.

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+        declVector.push_back((NamedDecl *) importedDecl);
+        llvm::ArrayRef < NamedDecl * > FoundDecls(declVector);
+        SetExternalVisibleDeclsForName(childCurrentDeclContext,
+                                       ((NamedDecl *) importedDecl)->getDeclName(),
+                                       FoundDecls);
+
+        /* And also put the Decl I found from the parent Interpreter
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: You don't seem to put a Decl into the map? I guess the comment is incorrect?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+      if (m_DeclName_map.find(childDeclName.getAsString()) != m_DeclName_map.end())
+        return true;
+
+      DeclContext::lookup_result lookup_result;
+      /* If we are not looking for this Name in the Translation Unit
+       * but instead inside a namespace, */
+      if (!childCurrentDeclContext->isTranslationUnit()) {
+        /* Search in the map of the stored Decl Contexts for this
+         * namespace. */
+        if (m_DeclContexts_map.find(childCurrentDeclContext) != m_DeclContexts_map.end()) {
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Can you store the result of `find()` and reuse it below, instead of calling `find()` again?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+                                 DeclarationName &parentDeclName) {
+
+      /* Check if we found this Name in the parent interpreter */
+      if (lookup_result.empty())
+        return false;
+
+      /* Prepare to import the Decl(Context)  we found in the
+       * child interpreter */
+      const FileSystemOptions systemOptions;
+      FileManager fm(systemOptions, nullptr);
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Why do you need to create your own `FileManager`? I'd assume that you should be using either the one of the parent or the child. But then again I don't know why the importer needs a FileManager. Could you add a comment?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+
+      /*****************************ASTImporter**********************************/
+      ASTImporter importer(to_ASTContext, fm, from_ASTContext, fm,
+        /*MinimalImport : ON*/ true);
+      /**************************************************************************/
+
+      /* If this Name we are looking for is for example a Namespace,
+       * then it is a Decl Context. */
+      if ((*lookup_result.data())->getKind() == Decl::Namespace) {
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: In general you want to `dyn_cast` the `Decl` to a `DeclContext`.

Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: I believe you need to check whether the lookup result is a single decl here...

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+        DeclContext *declContextToImport =
+          llvm::cast<DeclContext>(*lookup_result.data());
+        ImportDeclContext(declContextToImport, importer, childDeclName,
+                          parentDeclName, childCurrentDeclContext);
+      } else { /* If this name is just a Decl. */
+        /* Check if we have more than one results, this means that it
+         * may be an overloaded function. */
+        if (lookup_result.size() > 1) {
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Can't you just use this branch also for the case where you only have a single Decl in the lookup result?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 33]
```diff
@@ -0,0 +1,189 @@
...
+          DeclContext::lookup_iterator I;
+          DeclContext::lookup_iterator E;
+          for (I = lookup_result.begin(), E = lookup_result.end(); I != E; ++I) {
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: It's nicer to declare at first use, i.e. `for (DeclContext::lookup_iterator I = lookup_result.begin(), E = lookup_result.end()`...

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 34]
```diff
@@ -0,0 +1,189 @@
...
+            NamedDecl *D = *I;
+            Decl *declToImport = llvm::cast<Decl>(D);
+            ImportDecl(declToImport, importer, childDeclName,
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: You can just pass `*I` here, no need to get a temporary `NamedDecl*` and then cast it to a `Decl*`!

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 35]
```diff
@@ -0,0 +1,189 @@
...
+       * Get the identifier info from the parent interpreter
+       * for this Name. */
+      llvm::StringRef name(childDeclName.getAsString());
+      IdentifierTable &parentIdentifierTable =
+        m_parent_Interp->getCI()->getASTContext().Idents;
+      IdentifierInfo &parentIdentifierInfo = parentIdentifierTable.get(name);
+      DeclarationName parentDeclName(&parentIdentifierInfo);
+
+      /* Check if we have already imported this Decl (Context). */
+      if (m_DeclName_map.find(childDeclName.getAsString()) != m_DeclName_map.end())
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Why do you first lookup the parent name and then check whether we need it? I'd reverse these two...

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 36]
```diff
@@ -0,0 +1,189 @@
...
+        return true;
+
+      DeclContext::lookup_result lookup_result;
+      /* If we are not looking for this Name in the Translation Unit
+       * but instead inside a namespace, */
+      if (!childCurrentDeclContext->isTranslationUnit()) {
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Why don't you just add the two `TranslationUnitDecl`s to the map, to get rid of this extra branch?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 37]
```diff
@@ -0,0 +1,189 @@
...
+        /* Search in the map of the stored Decl Contexts for this
+         * namespace. */
+        if (m_DeclContexts_map.find(childCurrentDeclContext) != m_DeclContexts_map.end()) {
+          /* If childCurrentDeclContext was found before and is already in the map,
+           * then do the lookup using the stored pointer. */
+          DeclContext *originalDeclContext =
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: If "original" is "parent" then I'd use "parent" for consistency.

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 38]
```diff
@@ -0,0 +1,189 @@
...
+            m_DeclContexts_map.find(childCurrentDeclContext)->second;
+
+          Decl *fromDeclContext = Decl::castFromDeclContext(originalDeclContext);
+          ASTContext &from_ASTContext = fromDeclContext->getASTContext();
+
+          Decl *toDeclContext = Decl::castFromDeclContext(childCurrentDeclContext);
+          ASTContext &to_ASTContext = toDeclContext->getASTContext();
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Can you add the parent and child `ASTContext`s as members (set during construction), instead of having to re-calculate them for each import?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 39]
```diff
@@ -0,0 +1,189 @@
...
+
+          lookup_result = originalDeclContext->lookup(parentDeclName);
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: What if the lookup failed?

File: interpreter/cling/lib/Utils/ASTImportSource.cpp [hunk 40]
```diff
@@ -0,0 +1,189 @@
...
+          /* Do the import */
+          if (Import(lookup_result, from_ASTContext, to_ASTContext,
+                     childCurrentDeclContext, childDeclName, parentDeclName))
+            return true;
+        }
+      } else { /* Otherwise search in the Translation Unit. */
+        ASTContext &from_ASTContext = m_parent_Interp->getCI()->getASTContext();
```
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: I assume that this will be collapsed with the other `if` case above - not reviewed.

File: interpreter/llvm/src/tools/clang/lib/Sema/SemaOpenMP.cpp [hunk 41]
```diff
@@ -105,7 +105,7 @@ class DSAStackTy {
           ConstructLoc(), OrderedRegion(false), InnerTeamsRegionLoc() {}
   };
 
-  typedef SmallVector<SharingMapTy, 64> StackTy;
+  typedef SmallVector<SharingMapTy, 4> StackTy;
```
Comment (karies) [inline interpreter/llvm/src/tools/clang/lib/Sema/SemaOpenMP.cpp]: Can you split this into a separate commit? I'd like to have documented (in its commit message) why you did it. And then we'll send it upstream.

File: interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp (hunk 19 above)
Comment (esakellari) [inline interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp]: Would this be ok?

// FIXME: In the case of the multiple interpreters (parent-child), 
 // the child interpreter doesn't include the runtime universe. 
 // The child interpreter will try to import this function from its 
 // parent interpreter, but it will fail, because this is a template function.
  // Once the import of template functions becomes supported by clang,
  // this check can go away.

File: interpreter/cling/lib/Utils/ASTImportSource.cpp (hunk 39 above)
Comment (esakellari) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: I do the check for the result of the lookup inside Import().

File: interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp (hunk 19 above)
Comment (karies) [inline interpreter/cling/lib/Interpreter/ValueExtractionSynthesizer.cpp]: Perfect! (But instead of "this check can go away" I'd rephrase as "this assert can be re-activated")

File: interpreter/cling/lib/Utils/ASTImportSource.cpp (hunk 39 above)
Comment (karies) [inline interpreter/cling/lib/Utils/ASTImportSource.cpp]: Yes, but you pass the `lookup_result` to `Import()` even if `lookup()` fails. Is that the correct behavior?

File: interpreter/cling/include/cling/Interpreter/Interpreter.h (hunk 21 above)
Comment (esakellari) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: I left this for the end. I guess that with the use of the private function I will replace the constructor delegation , right?

**Comment by karies** [comment]
Moved to #149
